fireant/slicer/widgets/reacttable.py

```diff
@@ -0,0 +1,324 @@
+            dimension = dimension_map[f_dimension_key]
+
+            columns.append({
+                'Header': getattr(dimension, 'label', dimension.key),
+                'accessor': f_dimension_key,
+            })
+
+        return columns
+
+    @staticmethod
+    def transform_metric_column_headers(data_frame, item_map, dimension_display_values):
+        """
+        Convert the metrics into ReactTable column header definitions. This includes any pivoted dimensions, which will
+        result in multiple rows of headers.
+
+        :param data_frame:
+        :param item_map:
+        :param dimension_display_values:
+        :return:
+        """
+
+        def get_header(column_value, f_dimension_key, is_totals):
+            if f_dimension_key == metrics_dimension_key or is_totals:
+                item = item_map[column_value]
+                return getattr(item, 'label', item.key)
+
+            else:
+                return getdeepattr(dimension_display_values, (f_dimension_key, column_value), column_value)
+
+        def _make_columns(df, previous_levels=()):
+            f_dimension_key = df.index.names[0]
+            groups = df.groupby(level=0) \
+                if isinstance(df.index, pd.MultiIndex) else \
+                [(level, None) for level in df.index]
+
+            columns = []
+            for column_value, group in groups:
+                is_totals = TOTALS_VALUE == column_value
+
+                column = {'Header': get_header(column_value, f_dimension_key, is_totals)}
+
+                levels = previous_levels + (column_value,)
+                if group is not None:
+                    next_level_df = group.reset_index(level=0, drop=True)
+                    column['columns'] = _make_columns(next_level_df, levels)
+
+                else:
+                    if hasattr(data_frame, 'name'):
+                        levels += (data_frame.name,)
+                    column['accessor'] = '.'.join(levels)
+
+                if is_totals:
+                    column['className'] = 'fireant-totals'
+
+                columns.append(column)
+
+            return columns
+
+        column_frame = data_frame.columns.to_frame()
+        return _make_columns(column_frame)
+
+    @staticmethod
+    def transform_data(data_frame, item_map, dimension_display_values):
+        """
+        WRITEME
+
+        :param data_frame:
+        :param item_map:
+        :param dimension_display_values:
+        :return:
+        """
+        result = []
+
+        for index, series in data_frame.iterrows():
+            if not isinstance(index, tuple):
+                index = (index,)
+
+            index = [x
+                     if x not in item_map
+                     else getattr(item_map[x], 'label', item_map[x].key)
+                     for x in index]
+
+            row = {}
+            for key, value in zip(data_frame.index.names, index):
+                if key is None:
+                    continue
```

Comment:
Some inline comments in these complex functions would be useful, otherwise it is hard to follow.